enhancements/installer/azure-support-known-cloud-environments.md

```diff
@@ -0,0 +1,231 @@
+
+
+	// networkResourceGroupName is the Resource Group for network resources like the Virtual Network and Subnets used by the cluster.
+	// If empty, the value is same as ResourceGroupName.
+	// +optional
+	NetworkResourceGroupName string `json:"networkResourceGroupName,omitempty"`
+    
+	// cloudName is the name of the Azure cloud environment which can be used to configure the Azure SDK
+	// with the appropiate Azure API endpoints.
+	// If empty, the value is equal to `AzurePublicCloud`.
+	// +optional
+	CloudName AzureCloudEnvironment `json:"cloudName,omitempty"`
+}
+
+// +kubebuilder:validation:Enum="";AzurePublicCloud;AzureUSGovernmentCloud;AzureChinaCloud;AzureGermanCloud
+type AzureCloudEnvironment string
+
+const (
+	AzurePublicCloud AzureCloudEnvironment = "AzurePublicCloud"
+	AzureUSGovernmentCloud AzureCloudEnvironment = "AzureUSGovernmentCloud"
+	AzureChinaCloud AzureCloudEnvironment = "AzureChinaCloud"
+	AzureGermanCloud AzureCloudEnvironment = "AzureGermanCloud"
+)
+```
+
+#### Kube Cloud Config Controller
+
+Since various kubernetes components like the kube-apiserver, kubelet (machine-config-operator), kube-controller-manager, cloud-controller-managers use the `.spec.cloudConfig` Config Map reference for cloud provider specific configurations, a new controller kube-cloud-config was introduced previously. The controller will perform the task of stitching configuration from Infrastructure object with the rest of the cloud config, such that all the kubernetes components can continue to directly consume a Config Map for configuration.
+
+The controller will use the `cloudName` from the Infrastructure object to set the [`cloud`](https://github.com/kubernetes/kubernetes/blob/89ba90573f163ee3452b526f30348a035d54e870/staging/src/k8s.io/legacy-cloud-providers/azure/auth/azure_auth.go#L45-L46) field for the Azure cloud config.
+
+If the user has already set a `cloud` value in the `.spec.cloudConfig` which doesn't match the value provided by the Infrastructure object, the controller should return error detailing the conflict in values.
+
+#### Installer
+
+##### Install Config
+
+The install config should allow the user to set the cloud name for Azure.
+
+```go
+// types/pkg/azure/platform.go
+type Platform struct {
+    ...
+	// cloudName is the name of the Azure cloud environment which can be used to configure the Azure SDK
+	// with the appropiate Azure API endpoints.
+	// If empty, the value is equal to `AzurePublicCloud`.
+	// +optional
+	CloudName AzureCloudEnvironment `json:"cloudName,omitempty"`
+}
+
+// +kubebuilder:validation:Enum="";AzurePublicCloud;AzureUSGovernmentCloud;AzureChinaCloud;AzureGermanCloud
+type AzureCloudEnvironment string
+
+const (
+	PublicAzureCloud AzureCloudEnvironment = "AzurePublicCloud"
+	USGovernmentAzureCloud AzureCloudEnvironment = "AzureUSGovernmentCloud"
+	AzureChinaCloud AzureCloudEnvironment = "AzureChinaCloud"
+	AzureGermanCloud AzureCloudEnvironment = "AzureGermanCloud"
```

Comment:
Is there are the reason why naming is not consistent? `{Type}AzureCloud vs Azure{Type}Cloud`?

Reply:
fixed to `Azure<Type>Cloud`